rs/replicated_state/tests/replicated_state.rs

```diff
@@ -710,6 +713,56 @@ fn time_out_messages_updates_subnet_input_schedules_correctly() {
     );
 }
 
+#[test]
+fn enforce_best_effort_message_limit() {
+    let mut fixture = ReplicatedStateFixture::with_canisters(&[CANISTER_ID, OTHER_CANISTER_ID]);
+
+    // Enqueue 4 best-effort incoming requests of increasing sizes for
+    // `CANISTER_ID`, `OTHER_CANISTER_ID` and `own_subnet_id` (i.e. subnet queues).
+    let own_subnet_id = CanisterId::from(fixture.state.metadata.own_subnet_id);
+    for (i, receiver) in [CANISTER_ID, OTHER_CANISTER_ID, CANISTER_ID, own_subnet_id]
+        .iter()
+        .enumerate()
+    {
+        let mut request = request_to(*receiver);
+        request.deadline = SOME_DEADLINE;
+        request.method_name = String::from_utf8(vec![b'x'; i * 10 + 1]).unwrap();
+        fixture.push_input(request.into()).unwrap();
+    }
+
+    assert_eq!(
+        (0, 0.into()),
+        fixture
+            .state
+            .enforce_best_effort_message_limit(u64::MAX.into()),
+    );
+
+    let best_effort_memory_usage = fixture.state.best_effort_message_memory_taken();
+    assert_eq!(
+        (0, 0.into()),
+        fixture
+            .state
+            .enforce_best_effort_message_limit(best_effort_memory_usage),
+    );
+
+    // Enforce a limit equal to the mean message size. This should shed everything
+    // but the first message we enqueued.
+    let mean_message_size = best_effort_memory_usage / 4;
+    assert_eq!(
+        (3, mean_message_size * 3 + 15.into()),
+        fixture
+            .state
+            .enforce_best_effort_message_limit(mean_message_size),
+    );
```

Comment:
Should we maybe call it again to make sure it doesn't remove anything the second time? I guess it's sort of like the call above with `best_effort_memory_usage` though.

Reply:
Done.